zkevm-circuits/src/util/cell_manager.rs

```diff
@@ -1,8 +1,286 @@
-trait CellManagerStrategy {
-    fn query_top_aligned_cell(&mut self, cell_type: CellType) -> Cell<F>;
-    fn query_cell(&mut self, cell_type: CellType) -> Cell<F>;
-    fn query_cells(&mut self, cell_type: CellType, count: usize) -> Vec<Cell<F>>;
-    fn columns(&self) -> &[CellColumn<F>];
-    fn get_width(&self) -> usize;
+use std::collections::HashMap;
+
+use eth_types::Field;
+use gadgets::util::Expr;
+use halo2_proofs::{
+    circuit::{AssignedCell, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, VirtualCells},
+    poly::Rotation,
+};
+
+use crate::{
+    evm_circuit::{table::Table, util::CachedRegion},
+    util::query_expression,
+};
+
+pub(crate) use super::cell_manager_strategy::*;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+/// CellType represent a category of cell (and column).
+pub(crate) enum CellType {
+    StoragePhase1,
+    StoragePhase2,
+    StoragePermutation,
+    LookupByte,
+    Lookup(Table),
+}
+
+impl CellType {
+    // The phase that given `Expression` becomes evaluateable.
+    pub(crate) fn expr_phase<F: Field>(expr: &Expression<F>) -> u8 {
+        use Expression::*;
+        match expr {
+            Challenge(challenge) => challenge.phase() + 1,
+            Advice(query) => query.phase(),
+            Constant(_) | Selector(_) | Fixed(_) | Instance(_) => 0,
+            Negated(a) | Expression::Scaled(a, _) => Self::expr_phase(a),
+            Sum(a, b) | Product(a, b) => std::cmp::max(Self::expr_phase(a), Self::expr_phase(b)),
+        }
+    }
+
+    /// Return the storage phase of phase.
+    pub(crate) fn storage_for_phase(phase: u8) -> CellType {
+        match phase {
+            0 => CellType::StoragePhase1,
+            1 => CellType::StoragePhase2,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Return the storage cell of the expression.
+    pub(crate) fn storage_for_expr<F: Field>(expr: &Expression<F>) -> CellType {
+        Self::storage_for_phase(Self::expr_phase::<F>(expr))
+    }
+}
+
+#[derive(Clone, Debug)]
+/// Cell is a (column, rotation) pair that has been placed and queried by the Cell Manager.
+pub(crate) struct Cell<F> {
+    pub(crate) expression: Expression<F>,
+    pub(crate) column_expression: Expression<F>,
+    pub(crate) column: Column<Advice>,
+    pub(crate) column_idx: usize,
+    pub(crate) rotation: usize,
+
+    // DEPRECATED: the same value as column_idx
+    pub(crate) cell_column_index: usize,
```

Comment:
explicitly mark it deprecated?
```suggestion
    #[deprecated(note="the same value as column_idx")]
    pub(crate) cell_column_index: usize,
```

Reply:
Done